File: app/src/main/java/com/edricchan/studybuddy/HelpActivity.java
package com.edricchan.studybuddy;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.transition.Fade;
import androidx.transition.TransitionManager;

import com.edricchan.studybuddy.adapter.HelpArticleAdapter;
import com.edricchan.studybuddy.interfaces.HelpArticle;
import com.edricchan.studybuddy.utils.DataUtil;
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity;
import com.google.gson.Gson;

import java.io.InputStreamReader;
import java.lang.ref.WeakReference;
import java.net.URL;
import java.util.Arrays;
import java.util.List;

public class HelpActivity extends AppCompatActivity {

	private RecyclerView featuredRecyclerView;
	private ConstraintLayout constraintLayout;
	private LinearLayout progressBarLayout;
	private SwipeRefreshLayout swipeRefreshLayout;
	private SharedPreferences preferences;
	private static final String TAG = SharedHelper.getTag(HelpActivity.class);

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
		setContentView(R.layout.activity_help);
		preferences = PreferenceManager.getDefaultSharedPreferences(this);
		featuredRecyclerView = findViewById(R.id.helpFeaturedRecyclerView);
		constraintLayout = findViewById(R.id.constraintLayout);
		progressBarLayout = findViewById(R.id.progressBar);
		swipeRefreshLayout = findViewById(R.id.swipeRefreshLayout);
		swipeRefreshLayout.setOnRefreshListener(this::loadFeaturedList);
		swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary);
		initRecyclerView();
		loadFeaturedList();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.menu_help, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case android.R.id.home:
				onBackPressed();
				return true;
			case R.id.action_refresh:
				swipeRefreshLayout.setRefreshing(true);
				loadFeaturedList();
				return true;
			case R.id.action_send_feedback:
				SharedHelper.launchUri(this, DataUtil.uriSendFeedback, preferences.getBoolean(DataUtil.prefUseCustomTabs, true));
				return true;
			case R.id.action_version:
				SharedHelper.showVersionDialog(this);
				return true;
			case R.id.action_licenses:
				Intent licensesIntent = new Intent(this, OssLicensesMenuActivity.class);
				startActivity(licensesIntent);
				return true;
			case R.id.action_source_code:
				SharedHelper.launchUri(this, DataUtil.uriSrcCode, preferences.getBoolean(DataUtil.prefUseCustomTabs, true));
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	/**
	 * Initialises the recycler view by calling the needed methods
	 */
	private void initRecyclerView() {
		featuredRecyclerView.setHasFixedSize(true);
		featuredRecyclerView.setItemAnimator(new DefaultItemAnimator());
		featuredRecyclerView.setLayoutManager(new LinearLayoutManager(this));
		featuredRecyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
	}

	/**
	 * Loads the featured items from the JSON file
	 */
	private void loadFeaturedList() {
		try {
			progressBarLayout.setVisibility(View.VISIBLE);
			featuredRecyclerView.setVisibility(View.GONE);
			new GetHelpArticlesAsyncTask(this).execute(new URL(DataUtil.urlHelpFeatured));
		} catch (Exception e) {
			Log.e(TAG, "An error occurred while attempting to parse the JSON:", e);
		}
	}

	private static class GetHelpArticlesAsyncTask extends AsyncTask<URL, Void, List<HelpArticle>> {

		// Weak reference to the current activity
		// See this great post on StackOverflow for more info (to prevent memory leaks): https://stackoverflow.com/a/46166223
		private WeakReference<HelpActivity> activityRef;

		GetHelpArticlesAsyncTask(HelpActivity activity) {
			this.activityRef = new WeakReference<>(activity);
		}

		@Override
		protected List<HelpArticle> doInBackground(URL... urls) {
			InputStreamReader reader = null;
			try {
				reader = new InputStreamReader(urls[0].openStream());
			} catch (Exception e) {
				Log.e(TAG, "An error occurred while attempting to parse the JSON file:", e);
			}
			if (reader != null) {
				HelpArticle[] helpArticles = new Gson().fromJson(reader, HelpArticle[].class);
				return Arrays.asList(helpArticles);
			}
			return null;
		}

		@Override
		protected void onPostExecute(List<HelpArticle> helpArticles) {
			super.onPostExecute(helpArticles);

			if (helpArticles != null) {
				// Try to get a reference to the activity
				HelpActivity activity = activityRef.get();
				if (activity == null || activity.isFinishing()) return;
				// Update the adapter
				HelpArticleAdapter adapter = new HelpArticleAdapter(helpArticles);
				adapter.setOnItemClickListener((article, position) -> SharedHelper.launchUri(activity, article.getArticleUri(), activity.preferences.getBoolean(DataUtil.prefUseCustomTabs, true)));
				activity.featuredRecyclerView.setAdapter(adapter);
				activity.featuredRecyclerView.setVisibility(View.VISIBLE);
				activity.swipeRefreshLayout.setRefreshing(false);
				TransitionManager.beginDelayedTransition(activity.constraintLayout, new Fade(Fade.IN));
				activity.progressBarLayout.setVisibility(View.GONE);
			}
		}
	}

}

File: app/src/main/java/com/edricchan/studybuddy/ui/modules/help/HelpActivity.kt
package com.edricchan.studybuddy.ui.modules.help

import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.core.view.isVisible
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.transition.Fade
import androidx.transition.TransitionManager
import com.edricchan.studybuddy.R
import com.edricchan.studybuddy.annotations.AppDeepLink
import com.edricchan.studybuddy.annotations.WebDeepLink
import com.edricchan.studybuddy.constants.Constants
import com.edricchan.studybuddy.databinding.ActivityHelpBinding
import com.edricchan.studybuddy.extensions.TAG
import com.edricchan.studybuddy.extensions.startActivity
import com.edricchan.studybuddy.ui.modules.base.BaseActivity
import com.edricchan.studybuddy.ui.modules.help.adapter.HelpArticleAdapter
import com.edricchan.studybuddy.utils.UiUtils
import com.edricchan.studybuddy.utils.WebUtils
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity
import com.google.android.material.color.MaterialColors
import com.google.android.material.divider.MaterialDividerItemDecoration
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch

@WebDeepLink(["/help"])
@AppDeepLink(["/help"])
@AndroidEntryPoint
class HelpActivity : BaseActivity() {
    private lateinit var preferences: SharedPreferences
    private lateinit var binding: ActivityHelpBinding
    private lateinit var adapter: HelpArticleAdapter
    private lateinit var webUtils: WebUtils

    private val viewModel: HelpViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding = ActivityHelpBinding.inflate(layoutInflater).also { setContentView(it.root) }

        preferences = PreferenceManager.getDefaultSharedPreferences(this)

        binding.swipeRefreshLayout.apply {
            setOnRefreshListener { loadFeaturedList() }
            setColorSchemeColors(MaterialColors.getColor(this, R.attr.colorPrimary))
        }

        webUtils = WebUtils.getInstance(this)

        initRecyclerView()
        initAdapter()
        loadFeaturedList()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_help, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val webUtils = WebUtils.getInstance(this)
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }

            R.id.action_refresh -> {
                binding.swipeRefreshLayout.isRefreshing = true
                loadFeaturedList()
                true
            }

            R.id.action_send_feedback -> {
                webUtils.launchUri(Constants.uriSendFeedback)
                true
            }

            R.id.action_version -> {
                UiUtils.getInstance(this).showVersionDialog()
                true
            }

            R.id.action_licenses -> {
                startActivity<OssLicensesMenuActivity>()
                true
            }

            R.id.action_source_code -> {
                webUtils.launchUri(Constants.uriSrcCode)
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    /**
     * Initialises the recycler view by calling the needed methods
     */
    private fun initRecyclerView() {
        binding.helpFeaturedRecyclerView.apply {
            setHasFixedSize(true)
            itemAnimator = DefaultItemAnimator()
            layoutManager = LinearLayoutManager(this@HelpActivity)
            addItemDecoration(
                MaterialDividerItemDecoration(
                    this@HelpActivity,
                    MaterialDividerItemDecoration.VERTICAL
                ).apply {
                    isLastItemDecorated = false
                }
            )
        }
    }

    private fun initAdapter() {
        adapter = HelpArticleAdapter().apply {
            setOnItemClickListener { article, _ ->
                article.uri?.let { webUtils.launchUri(it) }
            }
        }.also { binding.helpFeaturedRecyclerView.adapter = it }

        lifecycleScope.launch {
            viewModel.helpArticles.flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
                .collect {
                    adapter.submitList(it)
                }
        }
    }

    /**
     * Loads the featured items from the JSON file
     */
    private fun loadFeaturedList() {
        try {
            binding.apply {
                progressBarLinearLayout.isVisible = true
                helpFeaturedRecyclerView.isVisible = false
            }
            // Load data
            lifecycleScope.launch {
                viewModel.refreshHelpArticles()
            }.invokeOnCompletion {
                if (it != null && it !is CancellationException) {
                    // An error occurred
                    Log.e(TAG, "Could not retrieve help articles:", it)
                }
                binding.apply {
                    helpFeaturedRecyclerView.isVisible = true
                    swipeRefreshLayout.isRefreshing = false
                    TransitionManager.beginDelayedTransition(constraintLayout, Fade(Fade.IN))
                    progressBarLinearLayout.isVisible = false
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "An error occurred while attempting to parse the JSON:", e)
        }
    }
}
